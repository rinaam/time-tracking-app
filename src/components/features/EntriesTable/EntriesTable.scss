@import '../../../styles/colors.scss';
@import '../../../styles/spacing.scss';

.empty-message {
    margin: $spacing-l 0;

    font: {
        size: 18;
        weight: 900;
    }
}