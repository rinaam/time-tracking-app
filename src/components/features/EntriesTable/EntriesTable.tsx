import React from 'react';
import { Person } from '../../../models/Person';
import { TimeEntryResponseModel } from '../../../models/TimeEntry';
import { TABLE_HEAD } from '../../../utils/constants';
import Button from '../../core/Button/Button';
import Table, { TableBodyField } from '../../shared/Table/Table';

interface EntriesTableProps {
  timeEntries?: TimeEntryResponseModel;
  person: Person;
  onDeleteEntry: (entryId: string) => void;
}

const EntriesTable: React.FC<EntriesTableProps> = ({ timeEntries, person, onDeleteEntry }) => {
  const mapTimeEntriesToTableBody = (entriesData?: TimeEntryResponseModel): TableBodyField[][] =>
    (entriesData?.data || []).map((entry) => {
      // this was very hard to figure out :(
      const serviceName =
        entriesData?.included.find((inc) => inc.id === entry.relationships.service.data.id)
          ?.attributes.name || '';

      const personName = `${person.attributes.first_name}  ${person.attributes.last_name}`;

      return [
        {
          component: <Button type="button" onClick={() => onDeleteEntry(entry.id)} text="delete" />,
          id: `delete-${entry.id}`,
        },
        { text: entry.attributes.date, id: `date-${entry.id}` },
        {
          text: serviceName,
          id: `service-${entry.id}`,
        },
        {
          text: personName,
          id: `person-${entry.id}`,
        },
        { text: entry.attributes.note, id: `note-${entry.id}` },
        { text: entry.attributes.time, id: `time-${entry.id}` },
      ];
    });

  if (!timeEntries) {
    return <div>loading time entries</div>;
  }

  return <Table tableHead={TABLE_HEAD} tableBody={mapTimeEntriesToTableBody(timeEntries)} />;
};

export default EntriesTable;
