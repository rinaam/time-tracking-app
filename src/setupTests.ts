// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import oranizationMembershipsResponse from './tests/fixtures/organizationMemberships';
import servicesFixture from './tests/fixtures/services';
import timeEntries from './tests/fixtures/timeEntries';

jest.useFakeTimers().setSystemTime(new Date('2022-05-21'));

// nice article https://kentcdodds.com/blog/stop-mocking-fetch

async function mockFetch(url: string) {
  switch (url) {
    case 'https://api.productive.io/api/v2/organization_memberships': {
      return {
        ok: true,
        status: 200,
        json: async () => oranizationMembershipsResponse,
      };
    }
    case 'https://api.productive.io/api/v2/services': {
      return {
        ok: true,
        status: 200,
        json: async () => servicesFixture,
      };
    }
    case 'https://api.productive.io/api/v2/time_entries?filter[person_id]=271497&filter[after]=2022-05-21&filter[before]=2022-05-22': {
      return {
        ok: true,
        status: 200,
        json: async () => timeEntries,
      };
    }

    default: {
      throw new Error(`Unhandled request: ${url}`);
    }
  }
}

// eslint-disable-next-line import/prefer-default-export
export const mockDeleteRequestRefetch = (timeEntryId: string) => async () => ({
  ok: true,
  status: 200,
  json: async () => ({
    ...timeEntries,
    data: timeEntries.data.filter((entry) => entry.id !== timeEntryId),
  }),
});

beforeAll(() => jest.spyOn(window, 'fetch'));
beforeEach(() => (window.fetch as jest.Mock).mockImplementation(mockFetch));
