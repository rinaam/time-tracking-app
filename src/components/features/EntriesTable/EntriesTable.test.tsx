import React from 'react';
import { render, waitFor, screen, within, fireEvent } from '@testing-library/react';
import EntriesTable from './EntriesTable';

import personResponse from '../../../tests/fixtures/person';
import { mockDeleteRequestRefetch } from '../../../setupTests';

describe('EntriesTable', () => {
  it('should render component and match snapshot', async () => {
    const { asFragment } = render(<EntriesTable person={personResponse} />);
    await waitFor(() => {
      expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getAllByText('Marina Savic').length).toEqual(2);
    });

    await waitFor(() => {
      expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should delete entry', async () => {
    const { asFragment } = render(<EntriesTable person={personResponse} />);

    await waitFor(() => {
      expect(screen.getByTestId('row-delete-14541918')).toBeInTheDocument();
    });

    const row = await screen.getByTestId('row-delete-14541918');
    const button = within(row).getByText('delete');

    fireEvent.click(button);

    (window.fetch as jest.Mock).mockImplementationOnce(mockDeleteRequestRefetch('14541918'));

    await waitFor(() => {
      expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.queryByTestId('row-delete-14541918')).not.toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
