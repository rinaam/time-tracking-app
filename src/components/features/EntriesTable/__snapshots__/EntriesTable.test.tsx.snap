// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntriesTable should render component and match snapshot when no time entries 1`] = `
<DocumentFragment>
  <p
    class="empty-message"
  >
    There is no tracked time for today
  </p>
</DocumentFragment>
`;

exports[`EntriesTable should render component and match snapshot when time entries are present 1`] = `
<DocumentFragment>
  <table
    class="table-wrapper"
  >
    <thead>
      <tr>
        <th>
          Actions
        </th>
        <th>
          Date
        </th>
        <th>
          Service
        </th>
        <th>
          Person
        </th>
        <th>
          Note
        </th>
        <th>
          Time
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        data-testid="row-delete-14541918"
      >
        <td>
          <button
            class="btn-wrapper"
            type="button"
          >
            delete
          </button>
        </td>
        <td>
          2022-05-21
        </td>
        <td>
          App design
        </td>
        <td>
          Marina  Savic
        </td>
        <td>
          <ul>
            <li>
              foo bar
            </li>
          </ul>
        </td>
        <td>
          100
        </td>
      </tr>
      <tr
        data-testid="row-delete-14541903"
      >
        <td>
          <button
            class="btn-wrapper"
            type="button"
          >
            delete
          </button>
        </td>
        <td>
          2022-05-21
        </td>
        <td>
          Design
        </td>
        <td>
          Marina  Savic
        </td>
        <td>
          <ul>
            <li>
              2
            </li>
          </ul>
        </td>
        <td>
          60
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
